File: crates/agent/src/faculty/space.rs
//! THE FIRST FACULTY, and the proof the seam is a generalisation rather than a
//! new thing beside the old one.
//!
//! Every field below is COPIED from what `components::SharedSpace` declared
//! when it was a component of its own, because the migration has to be inert:
//! the same id, the same slot, the same intent sentence and the same stability
//! render the same section, and `tests/prompt.rs` is the byte-for-byte proof.
//! The tools are what `subagent::with_the_space` returned — the space's three
//! and the workspace's set, which arrive together because the folder is the
//! space's (ADR-006, default deny: no space, no workspace).

use context::{Slot, Stability};

use super::Faculty;
use crate::components::Block;

/// The faculty's name, which is also its block's id and the key a host writes
/// its rendered parts under in `AgentState.senses`. One string, three jobs, so
/// they cannot drift apart.
pub const SPACE: &str = "space";

/// The one block this faculty contributes.
///
/// `SemiStatic`, and slotted ahead of the clock for that reason: a group's
/// facts change on the scale of a session, not of a turn, so the block belongs
/// inside the cacheable head — above the one section that can never be cached
/// rather than behind it.
const BLOCK: Block = Block {
    id: SPACE,
    slot: Slot::SPACE,
    intent: "The folder this group shares, what it has settled, what it has posted.",
    stability: Stability::SemiStatic,
};

/// The space faculty: what naming a space brings with it.
pub(super) fn faculty() -> Faculty {
    Faculty {
        name: SPACE,
        tools: [crate::space::space_tools(), crate::workspace::workspace_tools()].concat(),
        blocks: vec![BLOCK],
    }
}

File: packages/agent/src/faculty/space.js
/**
 * THE SPACE FACULTY — the folder a group shares, what it has settled, and what
 * it has posted. It is the proof the seam is a GENERALISATION rather than a new
 * thing beside the old one: `space:` in an agent file used to attach these
 * three tools and this one block directly, and now it declares a faculty that
 * does it, so a second faculty is a row in a table rather than a mechanism.
 *
 * The workspace tools the Rust attached alongside are NOT here. They are B30's
 * eleven descriptors and this build has none of them; attaching a name to a
 * tool that does not exist would put `write_file` in a prompt and a refusal at
 * the end of the round.
 * @module
 */

import { SLOT } from '@harness/context'
import { arg, tool } from '../tools.js'

/** The faculty's name, its block's id, and the key a host writes its rendered parts under. */
export const SPACE = 'space'

/** @type {import('./index.js').Faculty} */
export const spaceFaculty = {
  name: SPACE,
  // `semi_static`, and slotted ahead of the clock: a group's facts change on
  // the scale of a session, not of a turn.
  block: {
    id: SPACE,
    slot: SLOT.SPACE,
    intent: 'The folder this group shares, what it has settled, what it has posted.',
    stability: 'semi_static',
  },
  tools: [
    tool({
      name: 'remember',
      description: 'Record a fact in the shared space, for every agent working here to see.',
      args: [arg('key', 'string', 'what the fact is about'), arg('value', 'string', 'the fact itself, in one line')],
      mutates: true,
      needs: 'space',
    }),
    tool({
      name: 'forget',
      description: 'Remove a fact from the shared space once it is no longer true.',
      args: [arg('key', 'string', 'the key of the fact to remove')],
      mutates: true,
      needs: 'space',
    }),
    tool({
      name: 'post_note',
      description: 'Leave a note for the other agents working in this space.',
      args: [arg('note', 'string', 'the note, in one line')],
      mutates: true,
      needs: 'space',
    }),
  ],
}
